namerd/storage/k8s/src/main/scala/io/buoyant/namerd/storage/experimental/K8sDtabStore.scala

```diff
@@ -0,0 +1,136 @@
+    val (stream, close) = watchApi.dtabs.watch(None, None, dtabList.apiVersion)
+    stream.foldLeft(initState) { (nsMap, watchEvent) =>
+      val newState: NsMap = watchEvent match {
+        case DtabWatch.Added(a) => nsMap + toDtabMap(a)
+        case DtabWatch.Modified(m) => nsMap + toDtabMap(m)
+        case DtabWatch.Deleted(d) => nsMap - name(d)
+        case DtabWatch.Error(e) =>
+          log.error("k8s watch error: %s", e)
+          nsMap
+      }
+      witness.notify(Return(newState))
+      newState
+    }
+  }
+
+  /** List all Dtab namespaces */
+  def list(): Future[Set[String]] = act.run.sample match {
+    case Activity.Ok(nses) => Future.value(nses.keySet)
+    case Activity.Pending => for {
+      tryNses: Try[NsMap] <- act.values.toFuture
+      nses <- Future.const(tryNses)
+    } yield nses.keySet
+    case Activity.Failed(ex) => Future.exception(ex)
+  }
+
+  private[this] def namedDtab(ns: String, dtab: FDtab, version: Option[String] = None): Dtab = Dtab(
+    dentries = dtab,
+    apiVersion = Some(api.version),
+    kind = Some("DTab"),
+    metadata = Some(ObjectMeta(
+      name = Some(ns),
+      resourceVersion = version
+    ))
+  )
+
+  /**
+   * Update an existing dtab.  Returns a DtabVersionMismatchException if the
+   * supplied version doesn't match the current version.
+   */
+  def update(ns: String, dtab: FDtab, version: Buf): Future[Unit] = {
+    val versionStr = version match {
+      case Buf.Utf8(str) => str
+      case _ => throw new IllegalArgumentException(s"invalid or non-UTF8 version string: ${version.toString}")
+    }
+
+    api.dtabs.named(ns).put(namedDtab(ns, dtab, Some(versionStr))).rescue {
+      case io.buoyant.k8s.Api.Conflict(_) => Future.exception(new DtabVersionMismatchException())
+    }.unit
+  }
+
+  /**
+   * Update an existing dtab regardless of the current version or create a new
+   * dtab if one doesn't already exist.
+   */
+  def put(ns: String, dtab: FDtab): Future[Unit] = {
+    api.dtabs.named(ns).put(namedDtab(ns, dtab)).rescue {
+      // In the event of a 404 on a PUT, we should POST instead, as the k8s API doesn't appear to support creating on PUT.
+      // Note that this is potentially racy if the namespace is created between the PUT and the POST, but it's likely
+      // safer to surface that error all the way to the user in that case than to attempt to recover beyond this point.
+      case io.buoyant.k8s.Api.NotFound(_) => create(ns, dtab)
+    }.unit
+  }
+
+  /** Watch a dtab and its version. */
+  def observe(ns: String): Activity[Option[VersionedDtab]] =
+    act.map(_.get(ns))
+
+  /**
+   * Create a new dtab.  Returns a DtabNamespaceAlreadyExistsException if a
+   * dtab already exists with that namespace.
+   */
+  def create(ns: String, dtab: FDtab): Future[Unit] = {
+    api.dtabs.post(namedDtab(ns, dtab)).rescue {
+      case io.buoyant.k8s.Api.Conflict(_) => Future.exception(new DtabNamespaceAlreadyExistsException(ns))
+    }.unit
+  }
```

Comment:
I mildly prefer `.rescue.unit` to `.unit.rescue` because the former guarantees that the result is `Future[Unit]` whereas the latter could be anything depending on what's in the rescue block, but I also don't think it's a big deal either way.